MigrationCommitTest.assignCompletelyLostPartitionsIsCalledNoExceptions

Failed in http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4612

```
java.lang.AssertionError: expected null, but was:<java.lang.Exception>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotNull(Assert.java:755)
	at org.junit.Assert.assertNull(Assert.java:737)
	at org.junit.Assert.assertNull(Assert.java:747)
	at com.hazelcast.internal.partition.impl.MigrationCommitTest.assignCompletelyLostPartitionsIsCalledNoExceptions(MigrationCommitTest.java:514)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)

Hiccups measured while running test 'assignCompletelyLostPartitionsIsCalledNoExceptions(com.hazelcast.internal.partition.impl.MigrationCommitTest):'
10:31:30, accumulated pauses: 1739 ms, max pause: 242 ms, pauses over 1000 ms: 0
10:31:35, accumulated pauses: 3912 ms, max pause: 811 ms, pauses over 1000 ms: 0
10:31:40, accumulated pauses: 967 ms, max pause: 716 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/MigrationCommitTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/impl/MigrationCommitTest.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.internal.partition.impl;
 
+import com.hazelcast.cluster.Address;
 import com.hazelcast.cluster.ClusterState;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.ListenerConfig;
@@ -26,9 +27,6 @@ import com.hazelcast.internal.partition.InternalPartitionService;
 import com.hazelcast.internal.partition.MigrationInfo;
 import com.hazelcast.internal.partition.impl.MigrationInterceptorTest.MigrationInterceptorImpl;
 import com.hazelcast.internal.partition.impl.MigrationInterceptorTest.MigrationProgressNotification;
-import com.hazelcast.logging.LogEvent;
-import com.hazelcast.logging.LogListener;
-import com.hazelcast.cluster.Address;
 import com.hazelcast.spi.properties.ClusterProperty;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
@@ -41,11 +39,15 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
 import java.util.logging.Level;
+import java.util.logging.LogRecord;
 
 import static com.hazelcast.internal.partition.MigrationInfo.MigrationStatus.SUCCESS;
 import static com.hazelcast.internal.partition.impl.MigrationInterceptorTest.MigrationProgressEvent.COMMIT;
@@ -497,21 +499,30 @@ public class MigrationCommitTest extends HazelcastTestSupport {
         warmUpPartitions(hz1, hz2);
         waitAllForSafeState(hz1, hz2);
 
-        final AtomicReference<Throwable> exceptionRef = new AtomicReference<Throwable>();
-        hz1.getLoggingService().addLogListener(Level.WARNING, new LogListener() {
-            @Override
-            public void log(LogEvent logEvent) {
-                if (logEvent.getLogRecord().getThrown() != null) {
-                    exceptionRef.set(logEvent.getLogRecord().getThrown());
-                }
+        final AtomicReference<Throwable> exceptionRef = new AtomicReference<>();
+        hz1.getLoggingService().addLogListener(Level.WARNING, event -> {
+            LogRecord log = event.getLogRecord();
+            // We want to ensure there's no unexpected exception caught by MigrationThread.
+            if (!MigrationThread.class.getName().equals(log.getLoggerName())) {
+                return;
             }
+            exceptionRef.compareAndSet(null, log.getThrown());
         });
 
         hz1.getCluster().changeClusterState(ClusterState.NO_MIGRATION);
         hz2.getLifecycleService().shutdown();
         waitAllForSafeState(hz1);
 
-        assertNull(exceptionRef.get());
+        Throwable t = exceptionRef.get();
+        Supplier<String> messageSupplier = () -> {
+            StringWriter sw = new StringWriter();
+            if (t != null) {
+                sw.write("Unexpected exception! Stacktrace: \n");
+                t.printStackTrace(new PrintWriter(sw));
+            }
+            return sw.toString();
+        };
+        assertNull(messageSupplier.get(), t);
     }
 
     private Config createConfig() {

